packages/use-query-params/src/index.js

```diff
@@ -0,0 +1,59 @@
+import { parse, stringify } from 'query-string'
+import { useCallback, useEffect, useState } from 'react'
+import { useHistory } from 'react-router-dom'
+
+const useQueryParams = () => {
+  const { location: { search, pathname }, replace } = useHistory()
+  const parseFormat = useCallback(
+    () =>
+      parse(search, {
+        parseNumbers: true,
+        parseBooleans: true,
+        arrayFormat: 'comma',
+      }),
+    [search],
+  )
+  const defaultFnUpdater = useCallback(
+    (currentQueryParams, updater) => ({
+      ...currentQueryParams,
+      ...updater,
+    }),
+    [],
+  )
+  const [state, setState] = useState(parseFormat())
+
+  const setQueryParams = useCallback(
+    updater => {
+      const currentQueryParams = parseFormat()
+      const params =
+        updater instanceof Function
+          ? updater(currentQueryParams)
+          : defaultFnUpdater(currentQueryParams, updater)
+
+      const stringifiedParams = stringify(params, {
+        arrayFormat: 'comma',
+        sort: (a, b) => a.localeCompare(b),
+      })
+
+      if (stringifiedParams !== search.replace('?', '')) {
+        replace(`${pathname}?${stringifiedParams}`)
+
+        setState(params)
+      }
+
+      return state
```

Comment:
Isn't this a bad practice ? If you want to read from the hook you will use the returned hook value no ?

```js
const [queryParams, setQueryParams] = useQueryParam()

// bad
const newQueryParams = setQueryParams({ foo: 'bar' })
console.log(newQueryParams)

// good
setQueryParams({ foo: 'bar' })
console.log(queryParams)
```

Reply:
Yes you totally right !